jupyterhub/handlers/base.py

```diff
@@ -1105,7 +1115,15 @@ def _fail_api_request(self, user):
                 url_parts = urlparse(target)
                 query_parts = parse_qs(url_parts.query)
                 query_parts['redirects'] = redirects + 1
-                url_parts = url_parts._replace(query=urlencode(query_parts))
+                # https://stackoverflow.com/a/21997659/9899076
+                # https://docs.python.org/3/library/urllib.parse.html#urllib.parse.parse_qs
+                url_parts = url_parts._replace(query=urlencode(query_parts, doseq=True))
+                # prevent redirect loop when user_path != server_name
+                if user_path != server_name:
```

Comment:
What's the case addressed by this? Can you make a test for it? I don't follow the `split("/")[:3]` part, and I'm not sure it's going to be correct with respect to trailing slashes and requests for `/user/foo/servername/notebooks/somenotebook.ipynb`

Reply:
Oh you're definitely right, the `split("/")[:3]` wouldn't be correct for opening specific files at all. That wasn't something that had occurred to me. So that's probably just wrong. 

The **TL;DR** is that with the code for this section (as it is in the current master), there would be a redirect loop (and no spawn whatsoever I think) when `user_path != server_name`. But for some reason it would work (no redirect loop + successful spawn) when `user_path == server_name`. 

Basically all of the changes in this part are probably bad, or at least unnecessary, *except* the additional `doseq=True` argument in `urlencode`.**END TL;DR**

A better solution would have been to figure out why everything worked (with the re-direct code as it is in master) when `user_path == server_name` and then make sure the same thing happened when `user_path != server_name`. But after several hours I gave up and went with the hacky solution of re-directing/changing the URL to `user_path == server_name` in the cases when that wasn't true originally, and then everything "worked".

I.e. I didn't actually fix the problem, I just forcibly converted all of the cases where the problem occurred into special instances of the nice case where it didn't occur.

As you already point out above though that isn't a good idea since it is functionality-diminishing, e.g. inasmuch as it makes it impossible to access individual files on a named server.

Also @rcthomas told me it was already enough for NERSC's use case if it worked only when `user_path == server_name`. (So why bother having an implementation for when `user_path != server_name`? Maybe instead just raise an exception that says it's supposed to fail?)

And in any case my implementation of the forcible conversion to `user_path == server_name` (namely the `split("/")[:3]` thing) is implemented in a stupid way. I tried writing a regex for it but gave up after a couple of hours of failing to do so. My idea had been to use the same capturing groups as in the Tornado handler to parse the string, but that apparently doesn't work. (Or if it can be made to work, only with a lot of difficulty, since according to StackOverflow doing that would violate the spirit or intent of regex or something.) So basically I just gave up and hard coded in the `split("/")[:3]` part (`['', 'hub', 'user', 'user_name' ]`) to implement the forced conversion. Which is obviously dumb since it hard-codes a lot of assumptions which aren't stated explicitly and which can't be expected to hold in all circumstances.

There is a little bit more to the code in this section than the stupid parts above though. One thing I noticed is that, during the re-direct loops, every time the page got re-directed, an extra pair of square brackets was added to the query argument. So `server_name` (which remember is the query argument) would start out as `test`, but by the end of the re-direct loop it would be `'['['['test']']']'` or `` `[`[`[`test`]`]`]` `` or something else terrible like that. It's a result of a quirk of `parse_qs` documented at the two webpages in the comments above the line

    url_parts = url_parts._replace(query=urlencode(query_parts, doseq=True))

in particular, the extra `doseq=True` argument to `urlencode` is the fix. This doesn't matter for the code as it is in master, since the relevant GET requests don't have string-valued query arguments. But for this PR, which does introduce string-valued query arguments to that request, it seems that omitting the extra `doseq=True` argument would lead to unintended bugs. In particular, the addition of the spurious square brackets seemed to be part of what was causing the re-direct loop (since there may have even been a re-direct loop even when `user_path == server_name` without the fix), although they weren't the entire source of the problem.